lib/widgets/compose_box.dart

```diff
@@ -287,6 +289,61 @@ class _ContentInput extends StatefulWidget {
 }
 
 class _ContentInputState extends State<_ContentInput> {
+  String? _prevText;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.controller.addListener(_contentChanged);
+    widget.focusNode.addListener(_focusChanged);
+  }
+
+  @override
+  void didUpdateWidget(covariant _ContentInput oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller.removeListener(_contentChanged);
+      oldWidget.focusNode.removeListener(_focusChanged);
+      widget.controller.addListener(_contentChanged);
+      widget.focusNode.addListener(_focusChanged);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.controller.removeListener(_contentChanged);
+    widget.focusNode.removeListener(_focusChanged);
+    super.dispose();
+  }
+
+  void _contentChanged() {
+    if (_prevText == widget.controller.text) {
```

Comment:
You can find a version(*) of `String.==` if you have your IDE navigate to `_StringBase` (it's in `bin/cache/pkg/sky_engine/lib/_internal/vm/lib/string_patch.dart` in your Flutter tree).

```dart
  @pragma("vm:exact-result-type", bool)
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other is String && this.length == other.length) {
      final len = this.length;
      for (int i = 0; i < len; i++) {
        if (this.codeUnitAt(i) != other.codeUnitAt(i)) {
          return false;
        }
      }
      return true;
    }
    return false;
  }
```

As you can see there, it has a fast path if the strings are identical, and another if they're different lengths, returning true and false respectively. So the linear-time comparison happens only if the new string is a different instance yet has the same length.

I think this is probably fine. If we discover it's a performance issue, the fix might be to only check `identical`, and if not identical then assume the content was actually edited.

(*) I'm not sure that's the actual version that gets run, because the two concrete implementations `_OneByteString` and `_TwoByteString` override it like so:
```dart
  @pragma("vm:recognized", "asm-intrinsic")
  @pragma("vm:exact-result-type", bool)
  // Intrinsic is more efficient than an inlined body even for the small
  // strings.
  @pragma('vm:never-inline')
  bool operator ==(Object other) {
    return super == other;
  }
```

The `asm-intrinsic` thing sounds like the Dart VM may replace calls to that method with calls to an implementation written in assembly. I'd expect the assembly implementation to have all the same algorithmic optimizations as the pure-Dart version, though — it'll just have better constant factors.

Reply:
Oh OK, interesting!

> So the linear-time comparison happens only if the new string is a different instance yet has the same length.

Yeah, this does seem rare here.